pkg/chains/storage/grafeas/grafeas_test.go

```diff
@@ -408,71 +591,126 @@ type mockGrafeasServer struct {
 }
 
 func (s *mockGrafeasServer) CreateNote(ctx context.Context, req *pb.CreateNoteRequest) (*pb.Note, error) {
-	noteID := fmt.Sprintf("%s/notes/%s", req.GetParent(), req.GetNoteId())
-	expectedResponse := req.GetNote()
-	if s.notes == nil {
-		s.notes = make(map[string]*pb.Note)
+	notePath := fmt.Sprintf("%s/notes/%s", req.GetParent(), req.GetNoteId())
+	noteRequested := req.GetNote()
+
+	if s.entries == nil {
+		s.entries = make(map[string]*noteOccurrences)
 	}
 
-	if _, exists := s.notes[noteID]; exists {
+	if _, ok := s.entries[notePath]; ok {
 		return nil, gstatus.Error(codes.AlreadyExists, "note ID already exists")
 	}
-	s.notes[noteID] = expectedResponse
-	return expectedResponse, nil
+
+	s.entries[notePath] = &noteOccurrences{
+		note: noteRequested,
+	}
+	return noteRequested, nil
+}
+
+func (s *mockGrafeasServer) ListNotes(ctx context.Context, req *pb.ListNotesRequest) (*pb.ListNotesResponse, error) {
+	notes := []*pb.Note{}
+	for _, n := range s.entries {
+		notes = append(notes, n.note)
+	}
+	return &pb.ListNotesResponse{Notes: notes}, nil
 }
 
 func (s *mockGrafeasServer) ListOccurrences(ctx context.Context, req *pb.ListOccurrencesRequest) (*pb.ListOccurrencesResponse, error) {
 	// to make sure the occurrences we get are in order.
-	sortedOccurrencesInServer := []*pb.Occurrence{}
-	keys := []string{}
-	for k := range s.occurences {
-		keys = append(keys, k)
+	allOccurrencesInServer := []*pb.Occurrence{}
+	for _, note := range s.entries {
+		for _, occ := range note.occurrences {
+			allOccurrencesInServer = append(allOccurrencesInServer, occ)
+		}
 	}
-	sort.Strings(keys)
-	for _, k := range keys {
-		sortedOccurrencesInServer = append(sortedOccurrencesInServer, s.occurences[k])
+
+	filteredOccs := s.getOccurrencesByFilter(req.GetFilter(), allOccurrencesInServer)
+	return &pb.ListOccurrencesResponse{Occurrences: filteredOccs}, nil
+}
+
+func (s *mockGrafeasServer) ListNoteOccurrences(ctx context.Context, req *pb.ListNoteOccurrencesRequest) (*pb.ListNoteOccurrencesResponse, error) {
+	noteName := req.Name
+	if _, ok := s.entries[noteName]; !ok {
+		return nil, nil
+	}
+
+	allOccurrences := []*pb.Occurrence{}
+	for _, o := range s.entries[noteName].occurrences {
+		allOccurrences = append(allOccurrences, o)
 	}
 
+	filteredOccs := s.getOccurrencesByFilter(req.GetFilter(), allOccurrences)
+
+	return &pb.ListNoteOccurrencesResponse{Occurrences: filteredOccs}, nil
+}
+
+func (s *mockGrafeasServer) getOccurrencesByFilter(filter string, occurrences []*pb.Occurrence) []*pb.Occurrence {
+	sort.Slice(occurrences, func(i, j int) bool {
+		return occurrences[i].ResourceUri+occurrences[i].NoteName < occurrences[j].ResourceUri+occurrences[j].NoteName
+	})
+
 	// if filter string is empty, the expected behaviour will be to return all.
-	if len(req.GetFilter()) == 0 {
-		return &pb.ListOccurrencesResponse{Occurrences: sortedOccurrencesInServer}, nil
+	if len(filter) == 0 {
+		return occurrences
```

Comment:
I'd move the sorting back up to ListOccurrences - there's no mention that `getOccurrencesByFilter` is also sorting, and this might cause issues if this is removed / refactored later if the next change doesn't realize this did more than just filter.